salt/cli/support/collector.py

```diff
@@ -0,0 +1,537 @@
+
+    def link(self, title, path):
+        """
+        Add a static file on the file system.
+
+        :param title:
+        :param path:
+        :return:
+        """
+        # The filehandler needs to be explicitly passed here, so PyLint needs to accept that.
+        # pylint: disable=W8470
+        if not isinstance(path, file):
+            path = salt.utils.files.fopen(path)
+        self.__current_section.append({title: path})
+        # pylint: enable=W8470
+
+
+class SaltSupport(salt.utils.parsers.SaltSupportOptionParser):
+    """
+    Class to run Salt Support subsystem.
+    """
+
+    RUNNER_TYPE = "run"
+    CALL_TYPE = "call"
+
+    def _setup_fun_config(self, fun_conf):
+        """
+        Setup function configuration.
+
+        :param conf:
+        :return:
+        """
+        conf = copy.deepcopy(self.config)
+        conf["file_client"] = "local"
+        conf["fun"] = ""
+        conf["arg"] = []
+        conf["kwarg"] = {}
+        conf["cache_jobs"] = False
+        conf["print_metadata"] = False
+        conf.update(fun_conf)
+        conf["fun"] = conf["fun"].split(":")[-1]  # Discard typing prefix
+
+        return conf
+
+    def _get_runner(self, conf):
+        """
+        Get & setup runner.
+
+        :param conf:
+        :return:
+        """
+        conf = self._setup_fun_config(copy.deepcopy(conf))
+        if not getattr(self, "_runner", None):
+            self._runner = salt.cli.support.localrunner.LocalRunner(conf)
+        else:
+            self._runner.opts = conf
+        return self._runner
+
+    def _get_caller(self, conf):
+        """
+        Get & setup caller from the factory.
+
+        :param conf:
+        :return:
+        """
+        conf = self._setup_fun_config(copy.deepcopy(conf))
+        if not getattr(self, "_caller", None):
+            self._caller = salt.cli.caller.Caller.factory(conf)
+        else:
+            self._caller.opts = conf
+        return self._caller
+
+    def _local_call(self, call_conf):
+        """
+        Execute local call
+        """
+        try:
+            ret = self._get_caller(call_conf).call()
+        except SystemExit:
+            ret = "Data is not available at this moment"
+            self.out.error(ret)
+        except Exception as ex:  # pylint: disable=broad-except
+            ret = "Unhandled exception occurred: {}".format(ex)
+            log.debug(ex, exc_info=True)
+            self.out.error(ret)
+
+        return ret
+
+    def _local_run(self, run_conf):
+        """
+        Execute local runner
+
+        :param run_conf:
+        :return:
+        """
+        try:
+            ret = self._get_runner(run_conf).run()
+        except SystemExit:
+            ret = "Runner is not available at this moment"
+            self.out.error(ret)
+        except Exception as ex:  # pylint: disable=broad-except
+            ret = "Unhandled exception occurred: {}".format(ex)
+            log.debug(ex, exc_info=True)
```

Comment:
ret? And not `ex`?

Reply:
The purpose is to collect all events so it just puts the error as well.

Reply:
Sure, but the above `ret` is not used. I guess it was supposed to be on the log call, of not, nuke it.

Reply:
Ah, I got you. Sorry.